docs/architecture/adr/0012-angular-filename-convention.md

```diff
@@ -45,16 +45,27 @@ At Bitwarden we also use a couple of more types:
 - `.request` - Api Request
 - `.response` - Api Response
 - `.type` - Enum
-- `.service.abstraction` - Abstract class for a service, used for DI, not all services needs an
-  abstract class
 
 The class names are expected to use the suffix as part of their class name as well. I.e. a service
 implementation will be named `FolderService`, a request model will be named `FolderRequest`.
 
-In the event a service can't be fully implemented, an abstract class is created with the
-`Abstraction` suffix. This typically happens if the Angular and Node implementations have to differ
-for one reason or another. Traditionally interfaces would be used, but a TypeScript interface cannot
-be used to wire up dependency injection in JavaScript.
+## Abstract & Default Implementations
+
+The Bitwarden clients codebase serves multiple clients, one of which is node based and doesn't use
+the Angular Dependency Injection. In order to make code useable in both Angular and non Angular
+based clients we generally use abstract classes to define the interface. Ideally we would use
+_interfaces_ but TypeScript interfaces cannot be used to wire up dependency injection in JavaScript.
+
+All consumers will use the abstract class as a parameter in their constructor which will be manually
+wired up in the CLI and use Angular dependency injection in the Angular clients. In the case a
+dependency is only used in the Angular client, the abstract class can be omitted and the
+implementation referenced directly using `@Injectable`.
+
+To avoid naming conflicts, it's forbidden to name classes with the same name as the abstract class.
+Depending on the class usage the following prefixes are allowed:
+
+- `Default`: Used for the default implementation of an abstract class.
+- `Web`, `Browser`, `Desktop` and `Cli` for platform specific implementations.
```

Comment:
🤔 Has repetition in the file path been considered? I'd think it's safe, and perhaps even desirable for search purposes, to omit the platform-specific prefix from a file (or most differentiators) when that information is already captured in the module name. It's like a DRY microcosm.

A toy example based on some of the things I've seen:

WebPasswordGeneratorComponent
"web/tools/generator/password.component.html"
vs
"web/tools/generator/web-password-generator.component.html"

Reply:
Not really. The occurrences where we require different services in clients tends to be fairly small and should really only be applicable to services.

The example seems to be an old component that uses inheritance? Following more modern composition it wouldn't require a client specific prefix.

Reply:
This ADR doesn't seem to limit itself to services and talks about domain objects and so forth, so I'm interpreting it as applying across all of the code.

As far as the convention goes, there's a tendency towards abstraction in tools code. Components increasingly have base classes that provide reactive bindings, with client-specific code containing little more than a `@Component()` that bridges the bindings to a client-specific template. Since the component library work is presently limited to the web environment, the future state of this convention is unclear. 

We want a single responsive component, but the patterns we'd use to create them are still nascent, when they're realized at all. What can be said is that this is often the latest code available. They're not old components; they're complex, self-contained smart components.